File: December - 29/KishoreMuruganantham_minimum_weight_cycle.swift
import Foundation
var lines = [String]()
while let line = readLine() { lines.append(line) }
if !lines.isEmpty {
    func parse(_ l: String) -> String {
        if let r = l.range(of: "=") { return String(l[r.upperBound...]) }
        return l
    }
    
    let nStr = parse(lines[0])
    let n = Int(nStr.trimmingCharacters(in: .whitespaces)) ?? 0
    
    var adj = [[(Int, Int)]](repeating: [], count: n)
    
    if lines.count >= 2 {
        var rest = lines.dropFirst().joined(separator: " ")
        if let r = rest.range(of: "=") { rest = String(rest[r.upperBound...]) }
        rest = rest.replacingOccurrences(of: "[", with: "")
        rest = rest.replacingOccurrences(of: "]", with: "")
        rest = rest.replacingOccurrences(of: " ", with: "") // remove spaces to be safe
        let vals = rest.split(separator: ",").compactMap { Int($0) }
        
        let numEdges = vals.count / 3
        for i in 0..<numEdges {
            let u = vals[i*3]
            let v = vals[i*3+1]
            let w = vals[i*3+2]
            adj[u].append((v, w))
            adj[v].append((u, w))
        }
    }
    
    var minCycle = Int.max
    
    for s in 0..<n {
        var dist = [Int](repeating: Int.max, count: n)
        var parent = [Int](repeating: -1, count: n)
        var visited = [Bool](repeating: false, count: n)
        
        dist[s] = 0
        
        for _ in 0..<n {
            var minD = Int.max
            var u = -1
            
            for i in 0..<n {
                if !visited[i] && dist[i] < minD {
                    minD = dist[i]
                    u = i
                }
            }
            
            if u == -1 { break }
            visited[u] = true
            
            for (v, wt) in adj[u] {
                if dist[u] != Int.max {
                    if dist[u] + wt < dist[v] {
                        dist[v] = dist[u] + wt
                        parent[v] = u
                    } else if visited[v] && v != parent[u] {
                         let cycleLen = dist[u] + dist[v] + wt
                         if cycleLen < minCycle { minCycle = cycleLen }
                    }
                }
            }
        }
    }
    
    if minCycle == Int.max {
        print("-1")
    } else {
        print(minCycle)
    }
}

File: December - 29/KishoreMuruganantham_minimum_weight_cycle.R
input <- file("stdin", "r")
lines <- readLines(input)

l1 <- sub(".*=\\s*", "", lines[1])
v <- as.integer(l1)

rest <- paste(lines[-1], collapse = " ")
rest <- sub(".*=\\s*", "", rest)
rest <- gsub("\\[|\\]|\\s+", "", rest)
vals <- as.integer(unlist(strsplit(rest, ",")))

adj <- vector("list", v)
if (length(vals) > 0) {
  num_e <- length(vals) / 3
  for (i in seq_len(num_e)) {
    base <- (i - 1) * 3
    u <- vals[base + 1] + 1
    w_node <- vals[base + 2] + 1
    wt <- vals[base + 3]
    
    adj[[u]] <- c(adj[[u]], list(c(w_node, wt)))
    adj[[w_node]] <- c(adj[[w_node]], list(c(u, wt)))
  }
}

min_cycle <- Inf

for (s in 1:v) {
  dist <- rep(Inf, v)
  parent <- rep(-1, v)
  dist[s] <- 0
  
  visited <- rep(FALSE, v)
  
  for (count in 1:v) {
    min_d <- Inf
    u <- -1
    
    for (i in 1:v) {
      if (!visited[i] && dist[i] < min_d) {
        min_d <- dist[i]
        u <- i
      }
    }
    
    if (u == -1 || min_d == Inf) break
    visited[u] <- TRUE
    
    nbs <- adj[[u]]
    if (length(nbs) > 0) {
      for (nb in nbs) {
        idx <- nb[1]
        wt <- nb[2]
        
        if (dist[u] + wt < dist[idx]) {
          dist[idx] <- dist[u] + wt
          parent[idx] <- u
        } else if (visited[idx] && idx != parent[u]) {
          res <- dist[u] + dist[idx] + wt
          if (res < min_cycle) min_cycle <- res
        }
      }
    }
  }
}

if (min_cycle == Inf) {
  cat("-1\n")
} else {
  cat(min_cycle, "\n", sep = "")
}
